### PR TITLE
riemann-mysql: switch to riemann-c-client and add TLS support

### PR DESCRIPTION
**Pull request by pyr**
This is part of our effort to be able to switch riemann to a TLS-only mode.
Instead of hand-crafting protobuf frames, rely on https://github.com/algernon/riemann-c-client which is now available in most distributions.

This follows similar work in https://github.com/exoscale/rpp-c and https://github.com/collectd/collectd.

### CODE REVIEW

File: riemann-mysql.c [hunk 1]
```diff
@@ -166,45 +148,50 @@ mysql_get_handle(struct mysql_handler *hdl)
...
-int
-mysql_gather(struct mysql_handler *hdl, struct mysql_stats *stats)
+riemann_event_t *
+mysql_gather(struct mysql_handler *hdl)
 {
-	int		 desclen, io, sql;
-	
+	int		     desclen, io, sql;
+	enum state	 state;
+	int		     has_metric;
```
Comment (TooAngel) [inline riemann-mysql.c]: nitpicking: Unify indentation

Comment (pyr) [inline riemann-mysql.c]: will do.

File: riemann-mysql.c [hunk 2]
```diff
@@ -74,36 +76,18 @@ enum state {
 struct mysql_handler {
 	enum status	 status;
 	MYSQL		*db;
-	char		 host[MAXLEN];
-	int		 port;
+	char		 host[HOST_NAME_MAX];
+	int		     port;
```
Comment (sfriesel) [inline riemann-mysql.c]: tabs after non-tabs :-1:

Comment (pyr) [inline riemann-mysql.c]: remnants from the past, adapted.

File: riemann-mysql.c [hunk 3]
```diff
@@ -479,50 +397,69 @@ main(int argc, char *argv[])
 				key = val;
 				val = val + strcspn(key, " \t\n") + 1;
 				key[strcspn(key, " \t\n")] = '\0';
-				tags[tagcount++] = strdup(key);
+				tags[tagcount] = strdup(key);
+                if (tags[tagcount] == NULL)
```
Comment (TooAngel) [inline riemann-mysql.c:403]: Maybe unify use of tabs and spaces

File: riemann-mysql.c [hunk 4]
```diff
@@ -130,404 +109,363 @@ int		 riemann_send(struct riemann_host *, struct riemann_msg *);
...
+    if (mysql_real_query(hdl->db, MYSQL_QUERY, strlen(MYSQL_QUERY))) {
+        log_warn("could not execute query: %s", mysql_error(hdl->db));
+        state = S_UNKNOWN;
+        strncpy(description, mysql_error(hdl->db), desclen);
+        goto out;
+    }
+
+    res = mysql_store_result(hdl->db);
+    if (res == NULL) {
+        log_warn("could store result: %s", mysql_error(hdl->db));
```
Comment (TooAngel) [inline riemann-mysql.c]: `could not` ?

File: riemann-mysql.c [hunk 5]
```diff
@@ -130,404 +109,363 @@ int		 riemann_send(struct riemann_host *, struct riemann_msg *);
...
+        state = S_CRITICAL;
+        strncpy(description, mysql_error(hdl->db), desclen);
+        goto out;
+    }
+
+    row = mysql_fetch_row(res);
+    if (row == NULL) {
+        log_warn("could store result: %s", mysql_error(hdl->db));
```
Comment (TooAngel) [inline riemann-mysql.c]: Maybe change to `could not fetch result` to distinguish with previous message

File: riemann-mysql.c [hunk 6]
```diff
@@ -130,404 +109,363 @@ int		 riemann_send(struct riemann_host *, struct riemann_msg *);
...
+        state = S_UNKNOWN;
+        strncpy(description, mysql_error(hdl->db), desclen);
+        mysql_free_result(res);
+        goto out;
+    }
+
+    if (mysql_num_fields(res) < 33) {
```
Comment (TooAngel) [inline riemann-mysql.c]: `33` looks like a magic number to me, better give it a name.

File: riemann-mysql.c [hunk 7]
```diff
@@ -458,9 +360,25 @@ main(int argc, char *argv[])
 		} else if (strcasecmp(key, "mysql_database") == 0) {
 			strncpy(hdl.dbname, val, strlen(val) + 1);
 		} else if (strcasecmp(key, "riemann_host") == 0) {
-			strncpy(rh.host, val, strlen(val) + 1);
+			strncpy(riemann_host, val, strlen(val) + 1);
```
Comment (sfriesel) [inline riemann-mysql.c]: this range check looks wrong

File: riemann-mysql.c [hunk 8]
```diff
@@ -458,9 +360,25 @@ main(int argc, char *argv[])
...
 		} else if (strcasecmp(key, "riemann_port") == 0) {
-			rh.port = atoi(val);
+			riemann_port = atoi(val);
+        } else if (strcasecmp(key, "riemann_cert") == 0) {
+            strncpy(riemann_cert, val, strlen(val) + 1);
```
Comment (sfriesel) [inline riemann-mysql.c:369]: all of them

File: riemann-mysql.c (hunk 7 above)
Comment (pyr) [inline riemann-mysql.c]: How do you mean?
Maybe I should explain why I'm not doing boundary checks here: I rely on the fact that I check the length of lines with `MAX_LEN`. So I know the size of the buffers I'm dealing with.

I can then safely copy to destination buffers (the `+ 1` makes room for the final NUL byte.

**Comment by TooAngel** [comment]
LGTM

I would squash the commits before merging, but it is also okay like this.

**Comment by pyr** [comment]
Ready for merging. @sfriesel @TooAngel

**Comment by TooAngel** [comment]
'Only those with write access to this repository can merge pull requests.'

**Comment by pyr** [comment]
@TooAngel fixed.

**Comment by TooAngel** [comment]
:+1: :cake: :accept:

**Pull request merged**